pbm/restore/restore.go

```diff
@@ -132,10 +133,64 @@ func (r *Restore) Snapshot(cmd pbm.RestoreCmd, opid pbm.OPID, l *log.Event) (err
 	return r.Done()
 }
 
+func (r *Restore) ReplayOplog(cmd pbm.ReplayCmd, opid pbm.OPID, l *log.Event) (err error) {
+	defer func() {
+		if err != nil {
+			if err2 := r.MarkFailed(err); err2 != nil {
+				l.Error("mark restore as failed `%v`: %v", err, err2)
+			}
+		}
+
+		r.Close()
+	}()
+
+	if err = r.init(cmd.Name, opid, l); err != nil {
+		return errors.Wrap(err, "init")
+	}
+
+	if !r.nodeInfo.IsPrimary {
+		return fmt.Errorf("%q is not primary", r.nodeInfo.SetName)
+	}
+
+	if err := r.cn.SetOplogTimestamps(r.name, cmd.Start, cmd.End); err != nil {
+		return errors.Wrap(err, "set oplog timestamps")
+	}
+
+	chunk, err := r.cn.PITRFindChunk(r.nodeInfo.SetName,
+		primitive.Timestamp{T: uint32(cmd.Start)})
+	if err != nil {
+		return errors.WithMessage(err, "find first chunk")
+	}
+
+	startTS := chunk.StartTS
+	endTS := primitive.Timestamp{T: uint32(cmd.End)}
+	if err = r.prepareChunks(startTS, endTS); err != nil {
+		return errors.Wrap(err, "verify oplog slices chain")
+	}
```

Comment:
it won't work.

Reply:
It won't work. For example, if we have oplog from 10:00 to 12:00 and replay from 10:00 to 11:00, the following condition could fail because the start time (provided by CLI argument) has conceptually `T: "10:00", I: 0` but the first chunk can be `T: "10:00", I: 1`.
https://github.com/percona/percona-backup-mongodb/blob/822aff99cb2ba591dbcb7b0b32bb466b45ac3cfa/pbm/restore/restore.go#L365-L370

Therefore I tried to find the exact time for the first chunk.

Reply:
Yep, you, 're right. But we should probably change the `pbm.PITRGetChunksSlice` instead of having to make this first chunk extraction. I'll take a look at that later